### PR TITLE
8268877: TextInputControlSkin: incorrect inputMethod event handler after switching skin

### PR DESCRIPTION
**Pull request by andy-goryachev-oracle**
Using new Skin.install() method to properly set and remove inputMethodRequests property in TextInputControl.

This avoids memory leaks resulting from skin change, as well as honors user-set properties when installing an uninstalling.

This PR depends on [JDK-8290844](https://bugs.openjdk.org/browse/JDK-8290844) Skin.install().

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8268877](https://bugs.openjdk.org/browse/JDK-8268877): TextInputControlSkin: incorrect inputMethod event handler after switching skin

### Reviewers
 * [Ajit Ghaisas](https://openjdk.org/census#aghaisas) (@aghaisas - **Reviewer**)
 * [Kevin Rushforth](https://openjdk.org/census#kcr) (@kevinrushforth - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx pull/903/head:pull/903` \
`$ git checkout pull/903`

Update a local copy of the PR: \
`$ git checkout pull/903` \
`$ git pull https://git.openjdk.org/jfx pull/903/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 903`

View PR using the GUI difftool: \
`$ git pr show -t 903`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/903.diff">https://git.openjdk.org/jfx/pull/903.diff</a>

</details>

### CODE REVIEW

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/TextInputControlSkin.java [hunk 1]
```diff
@@ -324,83 +326,91 @@ protected boolean computeValue() {
...
+    @Override
+    public void install() {
+        super.install();
 
-        // FIXME: JDK-8268877 - incorrectly wired handler on replacing skin
+        TextInputControl control = getSkinnable();
         if (control.getOnInputMethodTextChanged() == null) {
             inputMethodTextChangedHandler = this::handleInputMethodEvent;
             control.setOnInputMethodTextChanged(inputMethodTextChangedHandler);
         }
```
Comment (kleopatra) [inline modules/javafx.controls/src/main/java/javafx/scene/control/skin/TextInputControlSkin.java:340]: for the record: this pattern is incorrect (no matter _where_ the skin implements it ;) - a skin should not set a singleton event handler on its control (see [JDK-8093590](https://bugs.openjdk.org/browse/JDK-8093590)).

Not sure how we should procede in this case: moving to install fixes the issue on face value (not overwriting a user-installed handler) but not the incorrect handler registration (should be replaced by an added listener as done in the fix of JDK-8093590) - the latter could be done in a follow-up issue, adding a code comment here?

Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/TextInputControlSkin.java:340]: Thank you.  You are right - we can use addEventHandler().

**Comment by kevinrushforth** [comment]
If this is changed to adding an `EventHandler` rather than calling `setOnInputMethodTextChanged`, do you still want / need to move the initialization logic to the `Skin::install` method? I presume you do, so we will need to highlight the potential behavioral incompatibility, in the API docs (maybe via an `@implNote`, since the only compatibility concern is with subclasses that make assumptions about undocumented behavior of their superclass) or a release note or both.

**Comment by andy-goryachev-oracle** [comment]
> do you still want / need to move the initialization logic to the `Skin::install` method?

good question.  In the case of `setOnInputMethodTextChanged` -> `control.addEventHandler(InputMethodEvent.INPUT_METHOD_TEXT_CHANGED` the code remains in the constructor, so no behavioral changes are expected.

But `control.setInputMethodRequests(inputMethodRequests);` must be done in `install()` since it's setting the property.

Another question is whether we should use ListenerHelper for `control.addEventHandler(InputMethodEvent.INPUT_METHOD_TEXT_CHANGED` , but that would move this PR into a draft stage until ListenerHelper is integrated.  Or perhaps it could be done as a part of [JDK-8296076](https://bugs.openjdk.org/browse/JDK-8296076).

**Comment by kevinrushforth** [comment]
> > do you still want / need to move the initialization logic to the `Skin::install` method?
> 
> good question. In the case of `setOnInputMethodTextChanged` -> `control.addEventHandler(InputMethodEvent.INPUT_METHOD_TEXT_CHANGED` the code remains in the constructor, so no behavioral changes are expected.
> 
> But `control.setInputMethodRequests(inputMethodRequests);` must be done in `install()` since it's setting the property.

In that case, depending on which direction you go, we may or may not need an impl note (and I wouldn't bother with a release note, since I think the compatibility risk is small enough anyway; I could be talked into a release note if others feel differently).

> Another question is whether we should use ListenerHelper for `control.addEventHandler(InputMethodEvent.INPUT_METHOD_TEXT_CHANGED` , but that would move this PR into a draft stage until ListenerHelper is integrated. Or perhaps it could be done as a part of [JDK-8296076](https://bugs.openjdk.org/browse/JDK-8296076).

Doing it as part of JDK-8296076 seems best.

**Comment by andy-goryachev-oracle** [comment]
I wonder if this even needs an implementation note, as it is a bug fix.
We did add one in [JDK-8292212](https://bugs.openjdk.org/browse/JDK-8292212) Skin.install(), should be sufficient.

**Comment by kevinrushforth** [comment]
Maybe not then.

**Review by aghaisas — APPROVED** [review]
Fix looks good to me!

**Review by kevinrushforth — APPROVED** [review]

**Pull request closed without merging**